Change DubboShutdownHook to singleton and process level

Chenjp opened the pull request:
When we register DubboShutdownHook, actions-on-exit are expected. So we should not disable the Hook.

## What is the purpose of the change
Change **DubboShutdownHook** to singleton, and make it as process level Hook. Fix #14429.

## Brief changelog
As declared, ShutdownHook should perform process-exit ops, familiar with dubbo engine. Move ```DubboShutdownHook#addShutdownHook``` back to ```DubboBootstrap```.

Update DubboShutdownHookTest to add more internal check and improve code coverage.

## Verifying this change


<!-- Follow this checklist to help us incorporate your contribution quickly and easily: -->

## Checklist
- [x] Make sure there is a [GitHub_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GitHub issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Check if is necessary to patch to Dubbo 3 if you are work on Dubbo 2.7
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Add some description to [dubbo-website](https://github.com/apache/dubbo-website) project if you are requesting to add a feature.
- [ ] GitHub Actions works fine on your own branch.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

Chenjp committed:
Remove unregister of ShutdownHook. fix #14429

When we register DubboShutdownHook, actions-on-exit are expected. So we should not disable the Hook.

AlbumenJ reviewed (changes requested):
We should remove the shutdown hook when Dubbo Application has been shutdowned by API.

AlbumenJ commented:
Ignore this exception would be better

Chenjp commented:
> We should remove the shutdown hook when Dubbo Application has been shutdowned by API.

@AlbumenJ I am confused by DubboShutdownHook. Per constructor, dubboShutdownHook is ApplicationModel level. But following lines shows that its may impact other application models of same fwk model.
```java
        // send readonly for shutdown hook
        List<GracefulShutdown> gracefulShutdowns =
                GracefulShutdown.getGracefulShutdowns(applicationModel.getFrameworkModel());
        for (GracefulShutdown gracefulShutdown : gracefulShutdowns) {
            gracefulShutdown.readonly();
        }
```

```java
// if all FrameworkModels are destroyed, clean global static resources, shutdown dubbo completely
```
According the above comment, dubbo server supports more than one FrameworkModel. 
TODO:
1. move ApplicationModel (and sub-models) cleanup functions back to ApplicationModel#onDestroy
2. Fwk model level cleanup - to FrameworkModel#onDestroy
3. Runtime hook on process exit, shutdown dubbo engine gracefully: DubboShutdownHook supposed to do.

AlbumenJ commented:
> * move ApplicationModel (and sub-models) cleanup functions back to ApplicationModel#onDestroy
> * Fwk model level cleanup - to FrameworkModel#onDestroy
> * Runtime hook on process exit, shutdown dubbo engine gracefully: DubboShutdownHook supposed to do.

Agree

Chenjp committed:
Change DubboShutdownHook to process level singleton

For those external managed moduleModel, checkAndWaitForTimeout.
Finally cleanup all resources arbitrarily when process exit.

Chenjp committed:
Update DubboShutdownHook.java

add more log.

Chenjp commented:
@

> > * move ApplicationModel (and sub-models) cleanup functions back to ApplicationModel#onDestroy
> > * Fwk model level cleanup - to FrameworkModel#onDestroy
> > * Runtime hook on process exit, shutdown dubbo engine gracefully: DubboShutdownHook supposed to do.
> 
> Agree

@AlbumenJ  The latest commit passes local mvn test, but receives github actions failure ("10 successful, 8 failing, and 1 skipped checks "). Anything wrong i miss?

Chenjp committed:
Sonar code smell enhancement

code smell

Chenjp committed:
Update DubboShutdownHook.java

Chenjp committed:
Merge branch 'apache:3.2' into 3.2-fix-14429-shutdownHook

Chenjp committed:
Update DubboShutdownHook.java

restore moduleModel external managed check to keep consistent with previous release.
add more log.

Chenjp committed:
Update DubboShutdownHookTest.java

FIX: testDestoryWithModuleManagedExternally - add log assertion to ensure.

Chenjp commented:
After my commits, following stack traces shows that recursive destroy occur (***framework->application->framework***). @AlbumenJ  may code refactoring is ***expected*** to match the productive-level design. 

We already define framework-application-module hierarchical relationship clearly in ```ScopeModel#internalId```comment section. And if we follow the relationship rule strictly, then ```XxxModel#destroyed.compareAndSet``` should have never been used.
```java
Request.setVersion(String) line: 95	
HeaderExchangeServer.sendChannelReadOnlyEvent() line: 134	
HeaderExchangeServer.close(int) line: 111	
DubboProtocol.destroy() line: 610	
ProtocolSecurityWrapper.destroy() line: 117	
ProtocolListenerWrapper.destroy() line: 99	
ProtocolFilterWrapper.destroy() line: 80	
ProtocolSerializationWrapper.destroy() line: 60	
InvokerCountWrapper.destroy() line: 55	
FrameworkModelCleaner.destroyProtocols(FrameworkModel) line: 68	
FrameworkModelCleaner.destroyFrameworkResources(FrameworkModel) line: 55	
FrameworkModelCleaner.onDestroy(FrameworkModel) line: 47	
FrameworkModelCleaner.onDestroy(ScopeModel) line: 1	
FrameworkModel(ScopeModel).notifyProtocolDestroy() line: 155	
FrameworkModel.tryDestroyProtocols() line: 280	
ApplicationModel.onDestroy() line: 159	
ApplicationModel(ScopeModel).destroy() line: 122	
FrameworkModel.onDestroy() line: 129	
FrameworkModel(ScopeModel).destroy() line: 122	
FrameworkModel.destroyAll() line: 210	
DubboShutdownHook.run() line: 128
```

Chenjp committed:
code refactoring

reduce code complexity and enhance test case.

Chenjp committed:
Move DubboShutdownHook to DubboBootstrap

The hook's a part of Bootstrap.

Chenjp committed:
Update DubboShutdownHook.java

Because DubboBootstrap does not follow it's original design - singleton, we have to introduce a key to prevent double-register.

Chenjp committed:
Update DubboShutdownHookTest.java

code smell update.

Chenjp committed:
add more tests

improve code coverage.
enhance log message check.

Chenjp committed:
Update DubboShutdownHookTest.java

format.

Chenjp commented:
> We should remove the shutdown hook when Dubbo Application has been shutdowned by API.

Chenjp commented:
> > * move ApplicationModel (and sub-models) cleanup functions back to ApplicationModel#onDestroy
> > * Fwk model level cleanup - to FrameworkModel#onDestroy
> > * Runtime hook on process exit, shutdown dubbo engine gracefully: DubboShutdownHook supposed to do.
> 
> Agree

Done. @AlbumenJ  PTAL

AlbumenJ commented on dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/bootstrap/DubboShutdownHook.java:
```diff
@@ -0,0 +1,157 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.dubbo.config.bootstrap;
```
Why move the package

Chenjp replied:
Currently, ```DubboShutdownHook``` is instantiated and registered in ```DubboBootstrap```. That's why make  ```DubboBootstrap``` invisible to others except ```DubboBootstrap```. We can consider it as a part of Bootstrap.

AlbumenJ replied:
`invisible`?

AlbumenJ replied:
In the same module and public method

Chenjp replied:
@AlbumenJ change DubboShutdownHook to a package type.
![2024-08-15 11_40_57-dubbo3 2 - dubbo-config-api_src_main_java_org_apache_dubbo_config_bootstrap_Dubb](https://github.com/user-attachments/assets/378e4ede-c678-4961-a06c-7e1b6aa3e1ba)

Chenjp replied:
@AlbumenJ Any new comments?

Chenjp replied:
> In the same module and public method

@AlbumenJ If any other concern, will recall the move-package-ops.

AlbumenJ replied:
Still don't know why

Chenjp committed:
DubboShutdownHook: enable two entrance support:DubboBootstrap and DubboSpringInitializer

DubboShutdownHook: enable two entrance support:DubboBootstrap and DubboSpringInitializer

Chenjp committed:
recall prev ops.

recall prev ops.

Chenjp committed:
DubboSpringInitializer update

DubboSpringInitializer update

Chenjp committed:
Update KeepRunningOnSpringClosedTest.java

Chenjp committed:
Revert "Update KeepRunningOnSpringClosedTest.java"

This reverts commit e30f38a2cda775315423d93d6c29013465eedb1f.

Chenjp committed:
Update FileSystemDynamicConfiguration.java

AlbumenJ commented on dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/DubboShutdownHook.java:
```diff
@@ -42,144 +42,133 @@ public class DubboShutdownHook extends Thread {
-        super("DubboShutdownHook");
-        this.applicationModel = applicationModel;
-        Assert.notNull(this.applicationModel, "ApplicationModel is null");
-        ignoreListenShutdownHook = Boolean.parseBoolean(
-                ConfigurationUtils.getProperty(applicationModel, CommonConstants.IGNORE_LISTEN_SHUTDOWN_HOOK));
```
Should support this

Chenjp replied:
@AlbumenJ because gracefulshutdown is composed with framework model, then wonder to know what will happen if two app model have different ignore_listen_shutdown configuration inner same framework model?
* Dubbo Application[1.0](DUBBO_INTERNAL_APPLICATION) -- those **internal app** should be skipped ?
* Dubbo Application[1.1](provider1)

AlbumenJ replied:
Yes, we should provide a way to disable `ShutdownHook`. In some special scenario, we cannot depend on it.

Chenjp replied:
Currently, the most valuable ops in the Hook are graceful shutdown command (just for dubbo protocol).

Chenjp committed:
register hook at application deployer and enable ignore listen shutdown hook

register hook at application deployer and enable ignore listen shutdown hook

Chenjp committed:
reduce cognitive complexity

Chenjp commented:
no more news.
